### PR TITLE
Support multiple Spring version integration tests

### PR DESCRIPTION
**Pull request by moarychan**
# Description

Support multiple Spring version integration tests based on configuration file `spring-cloud-azure-supported-spring.json`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/pipeline/tests-supported-spring-versions-template.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      MatrixConfigs: ${{ parameters.MatrixConfigs }}
+      PreGenerationSteps:
+        - script: |
+            python -m pip install termcolor
+          displayName: 'Install python module'
+        - script: |
+            python ./sdk/spring/scripts/compatibility_update_supported_version_matrix_json.py -mcp ${{ parameters.MatrixConfigs[0].Path }}
+          displayName: 'Update supported Spring versions'
+      MatrixFilters:
+        - JavaTestVersion=^(?!1.8|1.11).*
```
Comment (Netyyyy) [inline sdk/spring/pipeline/tests-supported-spring-versions-template.yml:76]: Does this available?

File: sdk/spring/pipeline/cosmos-integration-matrix.json [hunk 2]
```diff
@@ -10,6 +13,9 @@
       }
     },
     "TestFromSource": true,
-    "JavaTestVersion": ["1.11", "1.17", "1.21"]
+    "JavaTestVersion": ["1.11", "1.17", "1.21"],
```
Comment (Netyyyy) [inline sdk/spring/pipeline/cosmos-integration-matrix.json:16]: only cosmos support 1.11, should we align this with ours?

**Review by saragluna — APPROVED** [review]

File: eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml [hunk 3]
```diff
@@ -44,6 +44,12 @@ parameters:
 - name: Location
   type: string
   default: ''
+- name: PreGenerationSteps
```
Comment (saragluna) [inline eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml:47]: What are these steps used for?

File: eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml [hunk 4]
```diff
@@ -117,6 +123,9 @@ parameters:
 - name: UseFederatedAuth
   type: boolean
   default: false
+- name: AdditionalSparseCheckoutPaths
```
Comment (saragluna) [inline eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml:126]: What's this for?

File: sdk/spring/pipeline/tests-supported-spring-versions-template.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      AdditionalSparseCheckoutPaths:
```
Comment (saragluna) [inline sdk/spring/pipeline/tests-supported-spring-versions-template.yml:77]: Why we need to put the spring folder in the additional path?

File: sdk/spring/pipeline/tests-supported-spring-versions-template.yml (hunk 1 above)
Comment (moarychan) [reply sdk/spring/pipeline/tests-supported-spring-versions-template.yml:76]: Yes

File: sdk/spring/pipeline/cosmos-integration-matrix.json (hunk 2 above)
Comment (moarychan) [reply sdk/spring/pipeline/cosmos-integration-matrix.json:16]: Let me remove it.

File: eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml (hunk 3 above)
Comment (moarychan) [reply eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml:47]: It's a prerequisite to get the dynamic configuration before generating jobs. It's used to update the spring supported versions for matrix config file, such as `sdk/spring/pipeline/supported-version-matrix.json`, and `sdk/spring/pipeline/cosmos-integration-matrix.json`.

File: sdk/spring/pipeline/tests-supported-spring-versions-template.yml (hunk 5 above)
Comment (moarychan) [reply sdk/spring/pipeline/tests-supported-spring-versions-template.yml:77]: Integration test needs to dynamically obtain the supported Spring version. The executed scripts and configuration files are in `sdk/spring/pipeline` and `sdk/spring/scropts`, so they also need to be checked out before creating a job. Here they are passed to the parent template as an additional sparse checkout path configuration.

File: eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml (hunk 4 above)
Comment (moarychan) [reply eng/pipelines/templates/stages/archetype-sdk-tests-isolated.yml:126]: Same as the explanation below

**Comment by moarychan** [comment]
Hi @JimSuplizio, @benbp, @hallipr, @weshaggard, can any of you please review this PR? Thanks in advance!

File: sdk/spring/pipeline/cosmos-integration-matrix.json (hunk 2 above)
Comment (moarychan) [reply sdk/spring/pipeline/cosmos-integration-matrix.json]: Removed via PR https://github.com/Azure/azure-sdk-for-java/pull/41227

**Comment by moarychan** [comment]
Hi @JimSuplizio, @benbp, @hallipr, can any of you please review this PR? Thanks in advance!

**Comment by JimSuplizio** [comment]
@moarychan, I'm adding @alzimmermsft in case he has any feedback. I am going to say that it's kind of hard to review a PR considering the state of the [java - spring - tests](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3981571&view=results) pipeline results are indicative of something catastrophic going on. I cannot approve these changes at this time.

**Comment by moarychan** [comment]
Now Spring tests run almost completely successfully, I will re-run the failed one, @JimSuplizio and @alzimmermsft, please help review. Many thanks!

**Review by JimSuplizio — APPROVED** [review]
I'm okay with the eng changes.

**Comment by benbp** [comment]
I will review this today

**Review by benbp — APPROVED** [review]

**Pull request merged**